### PR TITLE
refactor: extract invite procedures into dedicated invite router

### LINKED ISSUE #2877: Extract invite procedures from organization router into dedicated invite router
## Context

Follow-up from #2872. The organization router still contains 6 invite-related procedures that delegate to `InviteService` and use `ctx.prisma` directly:

- `createInvites`
- `deleteInvite`
- `getOrganizationPendingInvites`
- `createInviteRequest`
- `approveInvite`
- `acceptInvite`

These were intentionally left in place during the org router refactoring because they already delegate to a proper service (`InviteService`). However, they belong in their own router for separation of concerns.

## Proposed approach

1. Create `inviteRouter` in `src/server/api/routers/invite.ts`
2. Move the 6 invite procedures into it
3. Merge the invite router into the app router
4. Organization router drops from ~1434 to ~850 lines

## Acceptance criteria

- Invite procedures live in their own router
- Organization router contains only org-specific procedures
- Existing integration tests pass unchanged
- No API contract changes

### PR DESCRIPTION
**Pull request by drewdrewthis**
## Why

Closes #2877

The `organization` tRPC router had grown to ~1500 lines with invite-related procedures tangled into organization concerns. This violated single-responsibility — invites are their own bounded concept with their own lifecycle (waiting approval → pending → accepted/expired).

## What changed

- **New `invite.ts` router** — every invite procedure moved from `organization.ts` (~770 lines); `organization.ts` trimmed from ~1500 to ~630 lines, now only handles org CRUD, members, teams, and billing.
- **`api.invite.*` namespace** — router registered in `root.ts`; all 6 frontend callers and 4 SubscriptionPage test suites updated.
- **`teamRoleInputSchema` extracted** to `routers/schemas/teamRole.ts` — breaks circular import between the two routers.
- **`getOrganizationPendingInvites` permission hardened** — was accidentally set to `organization:view` in the new router; corrected to `organization:manage` to match the intentional tightening from `eadd6e38f` (pending invites expose admin intent — who's being added, with what role, to which teams; member read is a leak).
- **RBAC coverage test updated** — `rbac.member-leak-coverage.integration.test.ts` now calls `invite.getOrganizationPendingInvites` (the moved procedure) to cover the same invariant.
- **Type predicates over non-null assertions** — replaced 3 `customRoleId!` with proper guards per project rules.

No other behavior changes — procedure signatures, validation, RBAC checks, and email side-effects are identical.

## Test plan

- All 22 invite integration tests pass against the new router (`invite.integration.test.ts`).
- All 4 SubscriptionPage test suites pass with updated mocks (79 tests).
- RBAC coverage test (`rbac.member-leak-coverage.integration.test.ts`) verifies `invite.getOrganizationPendingInvites` returns UNAUTHORIZED for member role.
- `pnpm typecheck` clean.
- CI: `langwatch-app-ci` green — typecheck, lint, build, unit tests, integration tests, e2e all pass.

## Anything surprising?

- The `getOrganizationPendingInvites` permission regression was a silent bug introduced during the extraction — this PR fixes it. The fix is confirmed by the RBAC integration test.
- Follow-up: invite procedure handlers are still fat (100+ line mutations with inlined logic). Extracting to `InviteService` is the natural next step but out of scope for this mechanical extraction.

---

## Rebase + fidelity restoration (2026-06-23, automated)

Rebased onto current `main` (branch was 80 commits behind). The mechanical rebase had a **single** conflict in `organization.ts` (main's `captureException(toError(err))` tweak vs the PR's deletion of the invite block) — resolved by taking the extraction.

After rebasing, the extracted `invite.ts` had drifted from main's *current* invite procedures (this branch predated main's invite-service refactor), so the "no behavior changes" claim above no longer held against current main. Re-synced to make the extraction genuinely behavior-preserving:

- **`acceptInvite`** now routes through `InviteService.applyInvite()` (instead of the stale inlined org/team membership creation) and restores the **Personal Workspace provisioning** block (`PersonalWorkspaceService.ensure(...)` wrapped in `captureException(toError(err))`) that had been silently dropped during extraction — now identical to main's `acceptInvite`. (This also aligns with @sergioestebance's resolved "use service/repo pattern here" note.)
- **`createInvites` / `createInviteRequest`** — restored the `kind: "custom"` filter (and the "user-assignable" comment) on custom-role validation, which was lost when the non-null assertions were refactored; without it a non-custom role id could pass validation.
- **`organization.ts`** — dropped the now-unused `PersonalWorkspaceService` import.

### Live-app proof — invite integration suite against real Postgres

```
$ pnpm test:pg-integration src/server/api/routers/__tests__/invite.integration.test.ts

 RUN  v4.1.4  .../langwatch

 Test Files  1 passed (1)
      Tests  22 passed (22)
   Duration  12.79s
```

The 22 cases exercise every extracted procedure end-to-end (`createInviteRequest`, `approveInvite`, `deleteInvite`, `getOrganizationPendingInvites`, `createInvites` batch, license-limit and email-failure paths) against a real database via the real Prisma client.

Also green locally: `acceptInvite` unit suite **6/6**, `pnpm typecheck` clean, `biome` clean on the changed router.

## Human verification

This is a pure refactor — invite procedures move from `api.organization.*` to `api.invite.*` with no behavior change. A human reviewer can verify:

1. **Members page** (`/settings/members`): invite flows work identically — create invite, view pending invites, accept/reject invite. The UI calls now route through `api.invite.*` but the user-visible behavior is unchanged.
2. **RBAC:** a member-role user attempting to call any invite-management endpoint (e.g. `getOrganizationPendingInvites`) receives UNAUTHORIZED — same restriction as before the refactor, now verified by the RBAC integration test.
3. **Namespace change is the only external delta:** any direct API calls using `api.organization.<invite-procedure>` need to update to `api.invite.<procedure>`. All 6 frontend callers are updated in this PR.

## How I can prove I was successful

The invariant is: every invite procedure behaves identically under the new `api.invite.*` namespace as it did under `api.organization.*`.

Evidence:
- 22 invite integration tests pass end-to-end against real Postgres via the new router (`invite.integration.test.ts`): `createInviteRequest`, `approveInvite`, `deleteInvite`, `getOrganizationPendingInvites`, `createInvites` batch, license-limit and email-failure paths all covered.
- RBAC invariant: `rbac.member-leak-coverage.integration.test.ts` confirms `invite.getOrganizationPendingInvites` returns UNAUTHORIZED for member role (the permission tightening from `eadd6e38f` is preserved).
- 79 SubscriptionPage tests pass with updated mocks.
- `pnpm typecheck` clean — no type regressions from the router move.
- CI green: `langwatch-app-ci` passes all gates (typecheck, lint, build, unit, integration, e2e).

**✅ Live-app visual proof — members UI served by the new `api.invite.*` router.**

Members page renders a pending invite via the new api.invite.getOrganizationPendingInvites router

Driven on the live golden instance running **this branch**. The Settings → Members page renders a pending invite in the **Invites** table. Captured network confirms the table is served by the **extracted** procedure `invite.getOrganizationPendingInvites` (batched tRPC query → HTTP 200), and the console carries **no** `No "query"-procedure on path "invite.getOrganizationPendingInvites"` error — exactly what a missing registration of the new router in `root.ts` would throw. Falsifiable: with the client on `api.invite.*` but the new router not yet registered, that read 404s server-side; once registered, the production members UI resolves through the extracted router unchanged. End-to-end create/approve/delete are covered by the 22 integration tests above against real Postgres.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

### Re-authored against current main (2026-07-07)

This branch was re-authored after a ~700-line modify/modify conflict in `organization.ts` (main evolved since the original branch point). Rather than resolve textually, the extraction was **re-applied against current main**:

- **Move-only, behavior-preserving.** The six invite procedures — `createInvites`, `deleteInvite`, `getOrganizationPendingInvites`, `createInviteRequest`, `approveInvite`, `acceptInvite` — were extracted **verbatim from main's current `organization.ts`** into `routers/invite.ts`. Bodies are byte-identical to main (modulo the standalone router wrapper / imports). No logic edits.
- The shared `teamRoleInputSchema` was extracted to `routers/schemas/teamRole.ts` (byte-identical to main's inline definition; it had no external importers).
- Router registered as `invite` in `root.ts`; all callers migrated `organization.*` → `invite.*` (public tRPC paths for these six procedures change accordingly — 45 call sites, plus renamed tests).
- `pnpm-lock.yaml` left at main's version (no dependency changes in a move-only refactor).
- **Note:** the earlier branch's review-driven cleanup that replaced two non-null assertions (`t.customRoleId!`) with type-guard filters was reverted to keep bodies byte-identical to main. That cleanup is a fine follow-up but is intentionally out of scope for this move-only extraction.

**Verification:** `pnpm typecheck` (tsgo) passes with 0 errors; `invite.acceptInvite.unit.test.ts` passes 6/6; grep confirms 0 stale `organization.*` callers of the moved procedures.

### CODE REVIEW

File: langwatch/src/server/api/routers/invite.ts [hunk 1]
```diff
@@ -0,0 +1,773 @@
...
+            teamIds: z.string().optional(), // Keep for backward compatibility
+            teams: z
+              .array(
+                z.object({
+                  teamId: z.string(),
+                  role: teamRoleInputSchema,
+                  customRoleId: z.string().optional(),
+                }),
+              )
+              .optional(),
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/invite.ts:48]: Out of scope for this PR — this is a pure move refactor with no behavior changes. The custom-role schema validation was identical in the original `organization.ts`. Filed as a follow-up concern but not addressing here to keep the refactor clean.

File: langwatch/src/server/api/routers/invite.ts [hunk 2]
```diff
@@ -0,0 +1,773 @@
...
+            })
+            .catch(captureException);
+
+          throw new TRPCError({
+            code: "FORBIDDEN",
+            message: error.message,
+          });
+        }
+        throw error;
+      }
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/invite.ts:116]: Out of scope — this PR moves invite procedures verbatim from `organization.ts` to `invite.ts` with no behavior changes. The deduplication logic (or lack thereof) is identical to what existed before. Good enhancement idea but belongs in a separate PR.

File: langwatch/src/server/api/routers/invite.ts [hunk 3]
```diff
@@ -0,0 +1,774 @@
...
+        await inviteService.checkLicenseLimits({
+          organizationId: input.organizationId,
+          newInvites: [{ role: invite.role, teams: teamAssignments }],
+          user: ctx.session.user,
+        });
+
+        return await inviteService.approveInvite({
+          inviteId: input.inviteId,
+          organizationId: input.organizationId,
+        });
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/invite.ts:592]: Out of scope — this PR is a verbatim move from `organization.ts`. The `approveInvite` enterprise-guard behavior is identical to what existed before. Filed as a follow-up concern for the future `InviteService` extraction noted in the PR description.

File: langwatch/src/server/api/routers/invite.ts [hunk 4]
```diff
@@ -0,0 +1,774 @@
...
+          message: "You must be signed in to accept the invite",
+        });
+      }
+
+      if (invite.status === "ACCEPTED") {
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Invite was already accepted",
+        });
+      }
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/invite.ts:665]: Out of scope — this PR is a verbatim move from `organization.ts`. The `acceptInvite` status check is identical to the pre-existing implementation. Worth a follow-up issue for the `WAITING_APPROVAL` bypass concern, but addressing it here would mix behavior change into a pure move refactor.

File: langwatch/src/server/api/routers/invite.ts [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+      const prisma = ctx.prisma;
+
+      const organization = await prisma.organization.findFirst({
+        where: {
+          id: input.organizationId,
+        },
+        include: {
+          members: true,
+        },
+      });
```
Comment (sergioestebance) [inline langwatch/src/server/api/routers/invite.ts:78]: we can take advantage to use service/repo pattern here

**Comment by drewdrewthis** [comment]
Marking this PR as blocked until it is hand-tested

**Comment by drewdrewthis** [comment]
⚠️ **preview boot timed out for `issue2877/extract-invite-procedures-from`**

URL: https://langwatch-pr-3008.boxd.sh  (last HTTP `502`)
SSH to debug: `ssh langwatch-pr-3008.boxd.sh` — then `docker ps` and `docker logs langwatch-app-1`

---
<sub>made with ❤️ by the <a href="https://boxd.sh">boxd.sh</a> team · boxd</sub>

**Comment by drewdrewthis** [comment]
**QA plan before merge** — testing strategy for the invite-router extraction (from planning/triage, for the orchardist).

CI is green, but this refactor can break invite flows silently, so functional verification + screenshots are required before merge.

**Approach — browser-qa against a forked golden:**
1. Fork or resume golden `drewdrewthis--langwatch-golden` (fallback: local `docker compose up -d --wait --build` → `localhost:5560`).
2. Checkout this branch on it.
3. Run the invite flow end-to-end: invite → accept → assert org membership + role + RBAC, plus the unhappy paths.
4. Screenshot each step → publish to `langwatch/pr-screenshots` → embed the raw URLs in this PR body.

**Highest-risk assertion:** the extraction must not have dropped AUTH/RBAC middleware on the moved procedures. Specifically confirm a member-role user CANNOT see pending invites — `getOrganizationPendingInvites` must return UNAUTHORIZED for members. This PR fixes a silent `organization:view`→`organization:manage` regression, and that is the exact invariant to verify live.

**DoD:** invite flow functionally verified + screenshots in this PR.

**Comment by drewdrewthis** [comment]
## Review verdict: READY

**Reviewed at:** `42bfd7a2` · **Run:** review (own-PR — rebase re-gate onto current main)

No blocking concerns. All 8 review threads resolved at this SHA, and the PR's **move-only refactor invariant is preserved** across the rebase from the prior clean base (`7dcde1013`) onto current `main`.

### Verification summary

| Check | Result | Evidence |
|-------|--------|----------|
| Rebase shape | ✅ | 1 commit on `origin/main`, linear (`merge-base --is-ancestor origin/main HEAD` OK); diff is the 24 invite-feature files |
| Sole conflict (`useAcceptInviteOnce.ts`) | ✅ `[hygiene-reviewer]` | Kept main's #5550 outcome-store fix (`onSuccess(data, variables)` → `recordInviteOutcome`), applied only the `api.organization.acceptInvite` → `api.invite.acceptInvite` rename; `acceptInvite` return shape byte-identical to main so `data.invite.organization.name` / `data.project?.slug` resolve; no stray old-path callers |
| Move-only invariant | ✅ `[researcher]` | All 6 extracted procedures' RBAC guards match `origin/main` exactly; `getAuditLogs` still `auditLog:view` (thread-7 fix intact); `organization.ts` = pure removal (1 import add, 748 del) |
| No dead imports | ✅ `[researcher]` | Neither `organization.ts` nor `invite.ts` imports `elasticMigrate`/`usageStatsQueue` (deleted from main in #4547) |
| Server router unchanged vs approved head | ✅ | `git diff 7dcde1013 HEAD -- invite.ts teamRole.ts errors.ts` = 0 lines |
| typecheck + tests | ✅ | `pnpm typecheck` clean; invite unit (6) + `useAcceptInviteOnce` hook integration (6) pass |
| CI @ HEAD | ✅ | 60 pass, 24 skipping, 0 fail |

### Non-blocking (New Issue)

- **`[researcher]`** Three `OnlineEvaluationDrawer*.test.tsx` files carry a Chakra-RadioCard test-selector fix (`getByLabelText` → `getByRole("radio", {name}).closest("label")`, ~119 lines) unrelated to the invite extraction. **Pre-existing** in the approved head (author's own `fix(tests): use getByRole …` + biome-format commits; **byte-identical across this rebase — not introduced here**), test-only, no runtime risk. Consider splitting into its own PR for commit-history hygiene; does not gate this PR.

---
_Verdict is prose, not a GitHub approval. Held review-ready — not merged._

**Review by 0xdeafcafe — DISMISSED** [review]

**Comment by 0xdeafcafe** [comment]
**`langwatch/e2e/auth-regression/iter47-invite-edge-cases.ts:302`** - URL still points at the old namespace. Test #5 asserts `trpcStatus === 401` for an unauthenticated POST, after this PR moves the procedure it'll see 404 (procedure-not-found) and fail. Not in CI so the merge stays green, but the auth-regression playbook quietly breaks for whoever runs it next.

One-liner fix: `/api/trpc/invite.acceptInvite?batch=1`.

(Posting as top-level since the file isn't in the PR diff and inline comments need a diff anchor.)

**Comment by 0xdeafcafe** [comment]
**`docs/ai-governance/members-and-invites.mdx:11` and `:61`** - still name `organization.createInvites`, which lives at `invite.createInvites` after this PR. The other three procedures on line 11 (`getAllOrganizationMembers`, `deleteMember`, `updateMemberRole`) stay on `organization.*`, only the invite ones moved.

(Top-level because the file isn't in the PR diff.)

**Comment by 0xdeafcafe** [comment]
**`specs/features/AUDIT_MANIFEST.md:356` and `:360`** - both reference `organization.createInvites` / `organization.createInviteRequest`, both moved to `invite.*` in this PR. Trivial `s///` pass.

(Top-level because the file isn't in the PR diff.)

File: langwatch/src/server/api/routers/__tests__/invite.acceptInvite.unit.test.ts [hunk 6]
```diff
@@ -120,7 +124,7 @@ describe("organization.acceptInvite", () => {
       project: { findFirst: vi.fn().mockResolvedValue(null) },
```
Comment (0xdeafcafe) [inline langwatch/src/server/api/routers/__tests__/invite.acceptInvite.unit.test.ts:124]: (Inline anchored here because line 105 isn't in any diff hunk - the actual finding is the `describe("organization.acceptInvite", ...)` at line 105 you can see in this hunk's context header.)

Sibling integration test got renamed (`organization.invites.integration.test.ts` -> `invite.integration.test.ts`), this one stayed. The file imports `inviteRouter` and only exercises the moved procedure, so both the filename and the describe label read as stale.

Worth a `git mv` to `invite.acceptInvite.unit.test.ts` and a matching describe update.

File: langwatch/src/hooks/useAcceptInviteOnce.ts [hunk 7]
```diff
@@ -22,7 +22,7 @@ export function _resetSubmittedInviteCodesForTests(): void {
 }
 
 type AcceptInviteMutation = ReturnType<
-  typeof api.organization.acceptInvite.useMutation
+  typeof api.invite.acceptInvite.useMutation
```
Comment (0xdeafcafe) [inline langwatch/src/hooks/useAcceptInviteOnce.ts:57]: (Inline anchored here because the actual finding line, 50, isn't in any diff hunk.)

The JSDoc at line 50 still names `organization.acceptInvite`, that's `invite.acceptInvite` after this PR moves it. Same drift in `langwatch/src/server/invites/errors.ts:7`.

File: langwatch/src/server/api/routers/__tests__/invite.acceptInvite.unit.test.ts (hunk 6 above)
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/__tests__/invite.acceptInvite.unit.test.ts:124]: Done in 3241ce80a — `git mv`'d to `invite.acceptInvite.unit.test.ts` and updated the top-level `describe` label to `invite.acceptInvite`. Thanks for the catch.

File: langwatch/src/hooks/useAcceptInviteOnce.ts (hunk 7 above)
Comment (drewdrewthis) [reply langwatch/src/hooks/useAcceptInviteOnce.ts:57]: Fixed in 3241ce80a — updated the JSDoc in `useAcceptInviteOnce.ts` and the comment in `server/invites/errors.ts` from `organization.acceptInvite` to `invite.acceptInvite`.

**Comment by langwatch-agent** [comment]
🐕 **PR Hound review brief** — assigned reviewer @sergioestebance

**Review mode:** Deep Review
**Scariest prod risk:** Incomplete `organization.* → invite.*` namespace migration — any surviving runtime caller of the old procedure names throws "no such procedure" at call time. The reviewer has already found stale references in an e2e auth-regression test, docs, and the audit manifest; those signal the grep was not exhaustive.

**Inspection targets** (max 3, each names a file or behavior):
- `[Must Check]` full-repo sweep for `organization.{createInvites,createInviteRequest,acceptInvite,getOrganizationPendingInvites}` — separate the runtime callers (e.g. `e2e/auth-regression/iter47-invite-edge-cases.ts:302` hits a real endpoint) from docs; a missed runtime caller is a silent break, not just stale text.
- `[Must Check]` `getOrganizationPendingInvites` permission raised `organization:view → organization:manage` — confirm `SubscriptionPage.tsx` seat-counting query still resolves for the roles that legitimately open that page; a viewer/member who could previously load it will now get UNAUTHORIZED.
- `[Ask Author]` `routers/schemas/teamRole.ts` (`teamRoleInputSchema` extraction to break the org↔invite circular import) — verify the two routers import the schema from the shared module and neither re-imports the other.

**Author questions** (optional, max 3):
- Did you grep for every old-namespace invite caller across langwatch/, agentic-e2e-tests/, docs/, and specs/? Reviewer comments suggest at least the e2e test and manifest were missed.
- The permission tightening to `organization:manage` — is that in-scope for a "no behavior change" refactor, or a deliberate fix riding along? Worth calling out explicitly so it gets RBAC scrutiny rather than being read as a pure move.

File: langwatch/src/server/api/routers/organization.ts [hunk 8]
```diff
@@ -1437,7 +693,7 @@ export const organizationRouter = createTRPCRouter({
         targetId: z.string().optional(),
       }),
     )
-    .use(checkOrganizationPermission("auditLog:view"))
+    .use(checkOrganizationPermission("organization:manage"))
```
Comment (drewdrewthis) [inline langwatch/src/server/api/routers/organization.ts]: **[correctness] `getAuditLogs` permission silently changed from `auditLog:view` → `organization:manage`**

This line was `checkOrganizationPermission("auditLog:view")` on main (confirmed at `organization.ts:1440` in main). The PR changes it to `organization:manage`, which **tightens** access — users who had `auditLog:view` but not `organization:manage` would be silently locked out of audit logs after merge.

The PR body explicitly claims *"No other behavior changes — procedure signatures, validation, RBAC checks, and email side-effects are identical."* This change contradicts that claim and is not described anywhere in the PR body or the rebase notes.

Likely a rebase conflict-resolution artifact in `organization.ts` (the deletions of invite procedures shift line numbers and can land context on the wrong hunk). Please either:
1. Revert to `"auditLog:view"` if this is unintentional, or
2. Document the intent (with rationale and a migration note if `auditLog:view` is being retired) if intentional.

**Action: Fix** (or at minimum Decide + document — the current state contradicts the PR's own correctness claim).

**Review by drewdrewthis — COMMENTED** [review]
Two-pass review of aa0687ea8. One blocking concern (undescribed permission change in organization.ts — see inline thread).

File: langwatch/src/server/api/routers/organization.ts (hunk 8 above)
Comment (drewdrewthis) [reply langwatch/src/server/api/routers/organization.ts]: Fixed in d8f751df7ca8b00d4bd04bd04dfd7823bfd4e130: restored `auditLog:view` exactly as main carries it (it had drifted to `organization:manage` during rebase). Also ran a full permission-string drift sweep across the rebased diff — no other drifted literals found; the extracted invite.ts procedures all carry their original main permissions, and pnpm-lock.yaml already matches main's copy (verified two-dot, no change needed).

**Review by 0xdeafcafe — DISMISSED** [review]

**Comment by drewdrewthis** [comment]
Slack PR review request posted.

**Review by 0xdeafcafe — CHANGES_REQUESTED** [review]
needs rebase due to restructure

**Comment by drewdrewthis** [comment]
[grinder] **Merge resumed after OOM incident — viability check, merge aborted**

Context: box OOM-killed the session mid-merge (21:20Z–23:42Z). Resumed the worktree, found MERGE_HEAD present with 22k staged paths and 16 unresolved conflicts (4 UU / 10 UD / 2 AU).

**Viability check — is the refactor still needed on main?**

Yes, the underlying refactor target still exists: invite procedures are still mixed into `organization.ts` on `origin/main` (1151 lines), not yet extracted:

```
$ git show origin/main:platform/app/src/server/api/routers/organization.ts | grep -n "createInvites\|deleteInvite\|getOrganizationPendingInvites\|createInviteRequest\|approveInvite\|acceptInvite"
589:  createInvites: protectedProcedure
711:  deleteInvite: protectedProcedure
744:  getOrganizationPendingInvites: protectedProcedure
761:  acceptInvite: protectedProcedure
```

No dedicated `invite.ts` router exists on main (`platform/app/src/server/api/routers/invite.ts` — fatal: does not exist in origin/main).

**But the merge itself is not salvageable as a mechanical conflict resolution.** The whole repo has been restructured since this branch last synced — the top-level `langwatch/` directory this PR was authored against no longer exists on main at all; everything moved to `platform/app/`:

```
$ git ls-tree -d origin/main -- langwatch    → (empty, gone)
$ git ls-tree -d origin/main -- platform/app → 040000 tree ... platform/app
```

The conflict set mixes stale `langwatch/...` paths with new `platform/app/...` paths (git's rename detection couldn't reconcile a whole-tree directory move against this branch's edits), which is exactly the "re-authored against current main" pattern this PR's own history already went through twice (2026-06-23, 2026-07-07 per PR body). Forcing a third textual merge through this would produce a merge, not a correct extraction — the PR needs re-authoring against current `platform/app/` paths, not conflict resolution.

**Action taken:** `git merge --abort`, worktree restored clean (HEAD `42bfd7a2` == `origin/issue2877/extract-invite-procedures-from`, `git status` clean). No force-push, no push at all — nothing was resolved.

**Recommendation:** re-author the extraction fresh against current `main` (as was done twice before), not another merge attempt. This is implementation work — routing to a coder, not resolving here.

Verified by:
  git show origin/main:platform/app/src/server/api/routers/organization.ts | grep -n ... → createInvites/deleteInvite/getOrganizationPendingInvites/acceptInvite all present, still inline
  git show origin/main:platform/app/src/server/api/routers/invite.ts → fatal: does not exist in origin/main
  git ls-tree -d origin/main -- langwatch → empty (directory gone from main)
  git status (post-abort) → clean, HEAD == origin/issue2877/extract-invite-procedures-from (42bfd7a2)

**Comment by drewdrewthis** [comment]
## Re-authored on the `platform/app` layout (2026-09-07, automated)

Main moved the app from `langwatch/` to `platform/app/` (ADR-076, 2bbeb7dd2e), so a plain rebase conflicted on every file. This branch was reset to `origin/main` and the extraction was re-applied against main's **current** invite code. Previous head `42bfd7a28ad990594188860c60b76a878ec4d8bd` is preserved for reference (not force-deleted from history; reachable by SHA).

**Moved from `organization` router to the new `invite` router (behavior-preserving, `api.invite.*`):**
- `createInvites`
- `deleteInvite`
- `resendInvite`
- `getOrganizationPendingInvites`
- `acceptInvite`

`teamRoleInputSchema` moved to `routers/schemas/teamRole.ts`. Tests renamed with `git mv`: `organization.invites.integration.test.ts` -> `invite.integration.test.ts`, `organization.acceptInvite.unit.test.ts` -> `invite.acceptInvite.unit.test.ts`. All callers, mocks, and `tsconfig.d11-scoped.json` updated.

**Dropped from the old PR (no longer applicable):**
- `createInviteRequest` / `approveInvite` extraction and the `agentic-e2e-tests/tests/members/steps.ts` edit — main replaced these with the `joinRequests` router.
- Unrelated formatting churn in `components/evaluations/__tests__/OnlineEvaluationDrawer*.test.tsx`.

@0xdeafcafe the rebase you requested is done; ready for another look once CI is green.
